src/Forms/GridField/GridFieldVersionTag.php

```diff
@@ -0,0 +1,186 @@
+
+class GridFieldVersionTag extends AbstractGridFieldComponent implements GridField_ColumnProvider
+{
+    /**
+     * @var string
+     */
+    protected $column = null;
+
+    /**
+     * Fields/columns to display version states. We can specifies more than one
+     * field but states only show in the first column found.
+     *
+     * @array
+     */
+    protected $versionedLabelFields = [];
+
+    public function __construct($versionedLabelFields = ['Name', 'Title'])
+    {
+        $this->setVersionedLabelFields($versionedLabelFields);
+    }
+
+    /**
+     * Column to decorate with version state
+     *
+     * @return string
+     */
+    public function getColumn()
+    {
+        return $this->column;
+    }
+
+    /**
+     * @param string $column
+     * @return VersionedGridFieldState
+     */
+    public function setColumn($column)
+    {
+        $this->column = $column;
+        return $this;
+    }
+
+    /**
+     * Search list for default column
+     *
+     * @return array
+     */
+    public function getVersionedLabelFields()
+    {
+        return $this->versionedLabelFields;
+    }
+
+    /**
+     * @param array $versionedLabelFields
+     * @return VersionedGridFieldState
+     */
+    public function setVersionedLabelFields($versionedLabelFields)
+    {
+        $this->versionedLabelFields = $versionedLabelFields;
+        return $this;
+    }
+
+    /**
+     * Modify the list of columns displayed in the table.
+     *
+     * @see {@link GridFieldDataColumns->getDisplayFields()}
+     * @see {@link GridFieldDataColumns}.
+     *
+     * @param GridField $gridField
+     * @param array $columns List reference of all column names.
+     */
+    public function augmentColumns($gridField, &$columns)
+    {
+        // Skip if not versioned, or column already set
+        if ($this->getColumn()) {
+            return;
+        }
+
+        $matchedVersionedFields = array_intersect(
+            $columns ?? [],
+            $this->versionedLabelFields
+        );
+
+        if (count($matchedVersionedFields ?? []) > 0) {
+            // Get first matched column
+            $this->setColumn(reset($matchedVersionedFields));
+        } elseif ($columns) {
+            // Use first column if none of preferred matches
+            $this->setColumn(reset($columns));
+        }
+    }
+
+    /**
+     * Names of all columns which are affected by this component.
+     *
+     * @param GridField $gridField
+     * @return array
+     */
+    public function getColumnsHandled($gridField)
+    {
+        return $this->getColumn() ? [$this->getColumn()] : [];
+    }
+
+    /**
+     * HTML for the column, content of the <td> element.
+     *
+     * @param GridField $gridField
+     * @param DataObject $record Record displayed in this row
+     * @param string $columnName
+     * @return string HTML for the column. Return NULL to skip.
+     */
+    public function getColumnContent($gridField, $record, $columnName)
```

Comment:
Strongly type params and return type, remove docblocks for param types unless there is useful information for a param/return type.  Do this for all methods in the PR

Reply:
There is a small problem, because `interface GridField_ColumnProvider` still doesn't support strong types, so I can update only few methods.